### PR TITLE
tutos: Inception of performance modes notebook

### PR DESCRIPTION
**Pull request by georgebisbas**
WIP: A PR that illustrates the impact of several optimization options to the generated code.
Opening as Draft PR to trigger discussion. 

After 1st discussion with @FabioLuporini 
* [x] Do another pass and improve the language. There's work to do in this direction.
* [x] Use an operator that results in producing tensor temporaries. Eq(u.forward, u.dx.dx) is the most trivial one that comes to my mind (@mloubout any other ideas?)
* [x] Iterate over the language again (== task one)
* [x] Extend the operator with two more equations, one that gets fused with the first one in the same loop nest, the other which gets scheduled to a different loop nest (when you get to this point, poke me. I'll be more than happy to provide some guidance)

### CODE REVIEW

File: examples/compiler/05_optimization_modes.ipynb [hunk 1]
```diff
@@ -36,7 +103,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "# This function will be used to print the difference between the generated code.\n",
+    "#TODO: George -- I guess we could (and should!) move this under examples/performance/utils.py ... \n",
```
Comment (georgebisbas) [inline examples/compiler/05_optimization_modes.ipynb:106]: Maybe we can exploit that with pytest at some point? I can think of several applications especially when asserting long strings ?

File: examples/README.md [hunk 2]
```diff
@@ -40,9 +40,9 @@ A set of more advanced examples are available in `seismic`:
 Further:
 
 * `mpi`: Jupyter notebooks explaining how MPI works in Devito.
-* `gpu`: Jupyter notebooks explaining how to run on GPUs with Devito.
 * `finance`: Jupyter notebooks with examples of applying Devito to partial differential equations with financial applications.
 * `misc`: Example operators outside the context of finite differences and
+* `performance`: Jupyter notebooks explaining the optimizations applied by Devito, the options available to steer the optimization process, how to run on GPUs, and much more.
   partial differential equations.
```
Comment (georgebisbas) [inline examples/README.md:46]: Is this line cut in the middle?

File: examples/performance/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+### Summary of optimization parameters
+
+* Levels
+  * `noop`: disable optimizations
+  * `advanced`: all optimizations
+  * `advanced-fsg`: alternative optimization pipeline
+
+* Options (type, default)
+  * Parallelism:
+    * `openmp` (boolean, False): disable/enable OpenMP parallelism
```
Comment (georgebisbas) [inline examples/performance/README.md:10]: maybe enable/disable? Convention.

File: examples/performance/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  * Blocking:
+    * `blockinner` (boolean, False): disable/enable loop blocking along innermost loop
```
Comment (georgebisbas) [inline examples/performance/README.md:12]: same

File: examples/performance/00_overview.ipynb [hunk 5]
```diff
@@ -0,0 +1,753 @@
...
+    "The following functions will be used throughout the notebook for various purposes."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#TODO: George -- I guess we could (and should!) move this under examples/performance/utils.py ... \n",
```
Comment (georgebisbas) [inline examples/performance/00_overview.ipynb:122]: YES, why not.

File: examples/performance/00_overview.ipynb [hunk 6]
```diff
@@ -0,0 +1,753 @@
...
+    "```\n",
+    "op1_omp.apply(..., autotune='aggressive')\n",
+    "```\n",
+    "\n",
+    "Loop blocking also provides two optimization options:\n",
+    "\n",
+    "* `blockinner={False, True}` -- to enable 3D (or any nD, n>2) blocking\n",
+    "* `blocklevels={int}` -- to enable hierarchical blocking, to exploit the cache hierarchy \n",
+    "\n",
+    "In the example below, we construct an `Operator` featuring six-dimensional loop blocking, where the first three loops represent outer blocks, whereas the second three loops represent inner blocks within an outer block."
```
Comment (georgebisbas) [inline examples/performance/00_overview.ipynb:305]: `featuring six-dimensional loop blocking`
Is this a good expression to use?

File: examples/performance/00_overview.ipynb [hunk 7]
```diff
@@ -0,0 +1,753 @@
...
+    "# print(_unidiff_output(str(op1_omp), str(op2_omp)))  # Uncomment to print out the diff only"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Code motion\n",
+    "\n",
+    "The `advanced` mode performs a code motion pass. In explicit PDE solvers, this is most commonly used to lift expensive time-invariant sub-expressions out of the inner loops. The pass is however quite general in that it is not restricted to the concept of time-invariance -- any sub-expression invariant with respect to a subset of `Dimension`s is a code motion candidate. In the example below, we see that `sin(f)` gets hoisted out of the inner loops since it is determined to be an expensive invariant sub-expression. In other words, the compiler trades the redundant computation of `sin(f)` for additional storage (the `r1[...]` array)."
```
Comment (georgebisbas) [inline examples/performance/00_overview.ipynb:344]: I think we move and cheap ones?

File: examples/performance/00_overview.ipynb [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+    "_print_kernel(op3_omp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Basic flop-reducing transformations\n",
+    "\n",
+    "Among the simpler flop-reducing transformations applied by the `advanved` mode we find:\n",
```
Comment (georgebisbas) [inline examples/performance/00_overview.ipynb:363]: advanced

File: examples/performance/00_overview.ipynb [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+   "source": [
+    "op5_omp = Operator(eq, opt=('cse', 'factorize', {'openmp': True}))\n",
+    "print(_unidiff_output(str(op5_omp), str(op4_omp)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There are no powers in this example, so the `opt-pows` pass has no effect."
```
Comment (georgebisbas) [inline examples/performance/00_overview.ipynb:403]: Why not adding some to the eq? TO show that as well?

File: examples/performance/00_overview.ipynb [hunk 10]
```diff
@@ -0,0 +1,753 @@
...
+   "source": [
+    "op14_omp = Operator(eq, opt=('advanced-fsg', {'openmp': True}))\n",
+    "print(op14_omp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `x` loop here is still shared by the two loop nests, by the `y` one isn't. Analogously, if we consider the alternative `eq` already used in `op7_b0_omp`, we get two completely separate, and therefore individually blocked, loop nests."
```
Comment (georgebisbas) [inline examples/performance/00_overview.ipynb:670]: but the y ?

File: examples/performance/00_overview.ipynb [hunk 11]
```diff
@@ -0,0 +1,753 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Performance optimization overview\n",
+    "\n",
+    "This tutorial illustrates the performance optimizations applied to the code generated by an `Operator`. As we shall see, most optimizations are automatically applied as they're known to systematically improve performance. Others, whose impact varies across different `Operator`s, are instead to be enabled through specific options.\n",
```
Comment (rhodrin) [inline examples/performance/00_overview.ipynb:9]: The tutorial seems to be 'explaining the optimizations flags that can be passed to `Operator`' more then 'illustrating the performance optimizations applied'?

Also, `Operator`s -> `Operator`'s

File: examples/README.md (hunk 2 above)
Comment (FabioLuporini) [reply examples/README.md:46]: good catch, dropping

File: examples/performance/00_overview.ipynb (hunk 8 above)
Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb:363]: fixing

File: examples/performance/00_overview.ipynb (hunk 10 above)
Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb:670]: fixing

File: examples/performance/00_overview.ipynb (hunk 11 above)
Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb:9]: >The tutorial seems to be 'explaining the optimizations flags that can be passed to Operator' more then 'illustrating the performance optimizations applied'?

perhaps both?

Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb:9]: > Also, Operators -> Operator's

fixing

**Comment by rhodrin** [comment]
Looks great. Just left some minor nitpicking comments - please ignore/incorporate as you wish.

File: examples/performance/00_overview.ipynb (hunk 5 above)
Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb:122]: ok, doing it

File: examples/performance/00_overview.ipynb (hunk 6 above)
Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb:305]: improving

File: examples/performance/00_overview.ipynb (hunk 9 above)
Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb:403]: good point, doing it

File: examples/performance/00_overview.ipynb (hunk 11 above)
Comment (FabioLuporini) [reply examples/performance/00_overview.ipynb]: > The tutorial seems to be 'explaining the optimizations flags that can be passed to Operator' more then 'illustrating the performance optimizations applied'?

addressed in the commit I'm about to push, thanks

File: examples/performance/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
+### Summary of optimization parameters
+
+* Levels
+  * `noop`: disable optimizations
+  * `advanced`: all optimizations
```
Comment (georgebisbas) [inline examples/performance/README.md:5]: enable all?

**Pull request merged**